src/main/java/com/puppycrawl/tools/checkstyle/GlobalStatefulModule.java

```diff
@@ -0,0 +1,47 @@
+////////////////////////////////////////////////////////////////////////////////
+// checkstyle: Checks Java source code for adherence to a set of rules.
+// Copyright (C) 2001-2017 the original author or authors.
+//
+// This library is free software; you can redistribute it and/or
+// modify it under the terms of the GNU Lesser General Public
+// License as published by the Free Software Foundation; either
+// version 2.1 of the License, or (at your option) any later version.
+//
+// This library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+// Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public
+// License along with this library; if not, write to the Free Software
+// Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
+////////////////////////////////////////////////////////////////////////////////
+
+package com.puppycrawl.tools.checkstyle;
+
+import java.lang.annotation.ElementType;
+import java.lang.annotation.Inherited;
+import java.lang.annotation.Retention;
+import java.lang.annotation.RetentionPolicy;
+import java.lang.annotation.Target;
+
+/**
+ * This annotation means that the module contains global context,
+ * which will be updated across the entire application's lifetime.
+ * across the entire application lifetime. This also means, that all files will be processed
+ * by the same module instance.
+ * This annotation should be used, if a module accumulates some information across the application
+ * lifetime, and processed only once at the end of application lifetime (however, the module still
+ * can produce some messages, while collecting information).
+ * The check methods and fields should be thread safe, because they may be accessed from separate
+ * threads at the same time.
+ * It is guaranteed that there will be exactly one module instance
+ * This is simular to multi-file validation, which checkstyle does not support fully yet.
```

Comment:
please add link to issue.

Reply:
@rnveach asked me to add this note. @rnveach could you please give me the issue link?

Reply:
it is https://github.com/checkstyle/checkstyle/issues/3540